st2actions/st2actions/container/base.py

```diff
@@ -127,6 +128,40 @@ def _do_run(self, runner):
                     (status, result, context) = runner.run(action_params)
                     result = jsonify.try_loads(result)
 
+            LOG.debug('Validating runner output: %s', runner.runner_type.output_schema)
+
+            if runner.runner_type.output_schema.get('unmodeled'):
+                runner.runner_type.output_schema.pop('unmodeled')
+                LOG.warn(
+                    """Deprecation Notice: This runner has previously had unmodeled
+                    output. In StackStorm 3.1 the output will be placed under the
+                    `output` key."""
+                )
+                runner_schema = {
+                    "additionalProperties": True
+                }
+            else:
+                runner_schema = {
+                    "type": "object",
+                    "properties": runner.runner_type.output_schema,
+                    "additionalProperties": False
+                }
+
+            LOG.info("Action Result: %s", result)
+
+            schema.validate(result, runner_schema, cls=schema.get_validator('custom'))
+
+            if runner.action.output_schema:
+                output_schema = copy.deepcopy(runner.runner_type.output_schema)
+                output_schema.update(runner.action.output_schema)
+                action_schema = {
+                    "type": "object",
+                    "properties": output_schema,
+                    "additionalProperties": False
+                }
+                LOG.debug('Validating action output: %s', action_schema)
+                schema.validate(result, action_schema, cls=schema.get_validator('custom'))
```

Comment:
Ah, I see - so we do two passes - first for runner schema and then for optional action schema.

Reply:
What if the runner is asynchronous like orquesta and mistral?  Where is the schema check when these runners post result back to st2? Orquesta handles and updates output for the live action and action execution at https://github.com/StackStorm/st2/blob/master/st2common/st2common/services/workflows.py#L829.  Mistral posts updates over API via https://github.com/StackStorm/st2mistral/blob/master/st2mistral/notifiers/stackstorm_notifier.py to PUT of action executions at https://github.com/StackStorm/st2/blob/master/st2api/st2api/controllers/v1/actionexecutions.py#L547.

Reply:
I believe the common denominator for both is update_status in action service.  So I would 1) refactor the output schema check into a method at action service 2) have the container base call this new method 3) add an output schema check in update_status of action service when result is updated 4) refactor update_execution_records in workflow service to also use the action_service.update_status instead of the update_liveaction_status and update_execution.